document/iterator.go

```diff
@@ -282,29 +283,41 @@ func (s Stream) GroupBy(groupFn func(d Document) (Value, error)) Stream {
 // Aggregate builds a list of aggregators for each group of documents and passes each document of the stream to them.
 func (s Stream) Aggregate(aggregatorBuilders ...AggregatorBuilder) Stream {
 	return NewStream(IteratorFunc(func(fn func(d Document) error) error {
-		aggregates := make(map[Value][]Aggregator)
-		var groups []Value
+		aggregates := make(map[string][]Aggregator)
+		var groupKeys []string
 
 		nullValue := NewNullValue()
 
+		var b bytes.Buffer
+
+		enc := NewValueEncoder(&b)
 		err := s.Iterate(func(d Document) error {
 			group := nullValue
 
 			if gd, ok := d.(*groupedDocument); ok {
 				group = gd.group
 			}
 
-			aggs, ok := aggregates[group]
+			b.Reset()
+
+			err := enc.Encode(group)
+			if err != nil {
+				return err
+			}
+
+			// using b.Bytes() and cast as a string is much faster than using b.String()
+			// because of compiler optimizations.
+			aggs, ok := aggregates[string(b.Bytes())]
```

Comment:
IIRC string([]byte(…)) conversion is surjective. Though I may be confusing rune to string conversions or something similar.

Also, that seems like a use case for a hashing algorithm. I think @tdakkota suggested adding `Hash` function for `document.Value`.

Reply:
I don't think hashing will be useful here as we need to encode the values anyway, hashing would mean doing the work of the map

Reply:
>Though I may be confusing rune to string conversions or something similar.

Just checked that, bytes to string conversion should be fine.